db/db_impl/db_impl_merge.cc

```diff
@@ -0,0 +1,368 @@
+        return Status::InvalidArgument(
+            "table_factory must be of the same type");
+      }
+    }
+  }
+  if (mopts.merge_memtable) {
+    if (rhs->total_log_size_ > 0) {
+      return Status::InvalidArgument("DB WAL is not empty");
+    }
+    if (rhs->write_buffer_manager_ != write_buffer_manager_) {
+      return Status::InvalidArgument(
+          "write_buffer_manager must be shared to merge memtable");
+    }
+  }
+  return Status::OK();
+}
+
+Status DBImpl::MergeDisjointInstances(const MergeInstanceOptions& merge_options,
+                                      const std::vector<DB*>& instances) {
+  Status s;
+  autovector<ColumnFamilyData*> this_cfds;
+  for (auto cfd : *versions_->GetColumnFamilySet()) {
+    assert(cfd != nullptr);
+    if (!cfd->IsDropped()) {
+      this_cfds.push_back(cfd);
+    }
+  }
+  const size_t num_cfs = this_cfds.size();
+
+  // # Sanity checks
+  // Check target instance (`this`).
+  if (two_write_queues_) {
+    return Status::NotSupported("target instance two_write_queues == true");
+  }
+  autovector<DBImpl*> db_impls;
+  autovector<DBImpl*> all_db_impls{this};
+  // A list of source db super versions grouped by cf. nullptr if the cf is
+  // missing.
+  autovector<autovector<SuperVersion*>> cf_db_super_versions;
+  std::shared_ptr<void> _defer(nullptr, [&](...) {
+    for (auto& db_super_versions : cf_db_super_versions) {
+      for (auto* super_version : db_super_versions) {
+        if (super_version != nullptr && super_version->Unref()) {
+          super_version->Cleanup();
+        }
+      }
+    }
+  });
+  // Check source instances.
+  for (size_t i = 0; i < instances.size(); i++) {
+    auto* db_impl = static_cast<DBImpl*>(instances[i]);
+    s = ValidateForMerge(merge_options, db_impl);
+    if (s.ok()) {
+      db_impls.push_back(db_impl);
+      all_db_impls.push_back(db_impl);
+    } else {
+      return s;
+    }
+  }
+
+  // # Internal key range check
+  assert(s.ok());
+  for (auto* this_cfd : this_cfds) {
+    auto& name = this_cfd->GetName();
+    auto* comparator = this_cfd->user_comparator();
+    using CfRange = std::pair<Slice, Slice>;
+    std::vector<CfRange> db_ranges;
+    std::vector<PinnableSlice> pinned_vals;  // hold temporary slice.
+    {
+      PinnableSlice smallest, largest;
+      bool found = false;
+      s = this_cfd->GetUserKeyRange(&smallest, &largest, &found);
+      if (!s.ok()) {
+        return s;
+      }
+      if (found) {
+        db_ranges.emplace_back(smallest, largest);
+        pinned_vals.push_back(std::move(smallest));
+        pinned_vals.push_back(std::move(largest));
+      }
+    }
+    for (auto* db : db_impls) {
+      auto cfd = db->versions_->GetColumnFamilySet()->GetColumnFamily(name);
+      if (cfd == nullptr || cfd->IsDropped()) {
+        continue;
+      }
+      PinnableSlice smallest, largest;
+      bool found = false;
+      s = cfd->GetUserKeyRange(&smallest, &largest, &found);
```

Comment:
Should we stop write and ref the superversion here? Otherwise changes may happen after check.